### PR TITLE
 engine(servermaster): init openapi framework and forward job master api

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflow/issues/5689

### What is changed and how it works?

This PR adds an initial openapi framework to servermaster.

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by amyangfei — COMMENTED** [review]
rest lgtm

**Review by amyangfei — APPROVED** [review]

**Comment by gozssky** [comment]
@maxshuang @liuzix PTAL, thanks.

**Comment by amyangfei** [comment]
make check failed
```go

[2022-06-20T03:33:39.153Z] engine/servermaster/openapi.go:188:19: Error return value of `c.AbortWithError` is not checked (errcheck)

[2022-06-20T03:33:39.153Z] 		c.AbortWithError(http.StatusBadRequest, errors.New("job id must not be empty"))

[2022-06-20T03:33:39.153Z] 		                ^

[2022-06-20T03:33:39.153Z] engine/servermaster/openapi.go:197:20: Error return value of `c.AbortWithError` is not checked (errcheck)

[2022-06-20T03:33:39.153Z] 			c.AbortWithError(http.StatusInternalServerError, errors.New(resp.Err.String()))

[2022-06-20T03:33:39.153Z] 			                ^

[2022-06-20T03:33:39.153Z] engine/servermaster/openapi.go:203:19: Error return value of `c.AbortWithError` is not checked (errcheck)

[2022-06-20T03:33:39.153Z] 		c.AbortWithError(http.StatusServiceUnavailable, errors.New("job is not online"))

[2022-06-20T03:33:39.153Z] 		                ^

[2022-06-20T03:33:39.153Z] engine/servermaster/openapi_test.go:53:9: indent-error-flow: if block ends with a return statement, so drop this else and outdent its block (revive)

[2022-06-20T03:33:39.153Z] 	} else {

[2022-06-20T03:33:39.153Z] 		return "", false

[2022-06-20T03:33:39.153Z] 	}

[2022-06-20T03:33:39.153Z] make: *** [check-static] Error 1

script returned exit code 2
```

File: engine/cmd/master/main.go [hunk 1]
```diff
@@ -58,6 +59,7 @@ func main() {
 		os.Exit(2)
 	}
 	version.LogVersionInfo()
+	gin.SetMode(gin.ReleaseMode)
```
Comment (maxshuang) [inline engine/cmd/master/main.go:62]: Do we need `DebugMode` and `TestMode` for future debug or test?

File: engine/docs/swagger/docs.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+                "produces": [
+                    "application/json"
+                ],
+                "tags": [
+                    "jobs"
+                ],
+                "summary": "Pause a job",
+                "parameters": [
+                    {
+                        "type": "string",
```
Comment (maxshuang) [inline engine/docs/swagger/docs.go:175]: We will remove the `pause` cmd from framework to user-defined cmd later.

File: engine/servermaster/http.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+	"github.com/pingcap/tiflow/engine/pkg/promutil"
+	"github.com/pingcap/tiflow/pkg/util"
+
+	// Used for OpenAPI online docs.
+	_ "github.com/pingcap/tiflow/engine/docs/swagger"
+)
+
+// RegisterRoutes create a router for OpenAPI
+func RegisterRoutes(router *gin.Engine, openapi *OpenAPI) {
```
Comment (maxshuang) [inline engine/servermaster/http.go:33]: Can we add a simple ut for the `RegisterRoutes `?

File: engine/cmd/master/main.go (hunk 1 above)
Comment (gozssky) [reply engine/cmd/master/main.go:62]: I think we can use environment variable to set debug or test mode and set release mode by default. I will fix later.

File: engine/servermaster/server.go [hunk 4]
```diff
@@ -575,11 +575,15 @@ func (s *Server) startGrpcSrv(ctx context.Context) (err error) {
 		close(registerDone)
 	}
 
+	router := gin.New()
+	openapi := NewOpenAPI(s.jobManager, s.executorManager)
```
[the remark this answers is not in the document]
Comment (maxshuang) [reply engine/servermaster/server.go:579]: Do you have some uts or integration tests for the openapi in servermaster?

File: engine/docs/swagger/docs.go (hunk 2 above)
Comment (maxshuang) [reply engine/docs/swagger/docs.go:175]: Can you add a TODO for me? I will remove it in my pr.

File: engine/servermaster/server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gozssky) [reply engine/servermaster/server.go]: I add a `ServerInfoProvider` to get `jobManager` and `executorManager` dynamically. PTAL again. Also cc @amyangfei .

[the remark this answers is not in the document]
Comment (maxshuang) [reply engine/servermaster/server.go]: LGTM

**Comment by maxshuang** [comment]
e2e fail. Need take a closer look at it. @gozssky

**Comment by amyangfei** [comment]
> e2e fail. Need take a closer look at it. @gozssky

It seems a framework bug, a job master reports its master (job manager) timeout very soon after a heartbeat ping-ping

```go
[2022/06/21 11:18:50.731 +00:00] [DEBUG] [master_client.go:231] ["sending heartbeat"] [worker=d197b818-5933-4bbf-bb93-89cb533b54fc]
[2022/06/21 11:18:50.731 +00:00] [INFO] [master_client.go:236] ["sending heartbeat success"] [worker=d197b818-5933-4bbf-bb93-89cb533b54fc] [master-id=dataflow-engine-job-manager]
[2022/06/21 11:18:50.730 +00:00] [WARN] [center.go:53] ["More than one error is received"] [error="[DFLOW:ErrWorkerSuicide]worker has committed suicide due to master(dataflow-engine-job-manager) having timed out"] [errorVerbose="[DFLOW:ErrWorkerSuicide]worker has committed suicide due to master(dataflow-engine-job-manager) having timed out\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/engine/framework.(*DefaultBaseWorker).runWatchDog\n\tgithub.com/pingcap/tiflow/engine/framework/worker.go:544\ngithub.com/pingcap/tiflow/engine/framework.(*DefaultBaseWorker).startBackgroundTasks.func2\n\tgithub.com/pingcap/tiflow/engine/framework/worker.go:501\nruntime.goexit\n\truntime/asm_amd64.s:1571"]
```

**Comment by gozssky** [comment]
> ```go
> [2022/06/21 11:18:47.728 +00:00] [INFO] [fake_master.go:582] ["new fake master"] [config="{\"job-name\":\"test-node-failure\",\"worker-count\":4,\"target-tick\":10000000,\"etcd-watch-enable\":true,\"etcd-endpoints\":[\"user-etcd-standalone:2379\"],\"etcd-watch-prefix\":\"/fake-job/test/\",\"etcd-start-revision\":0,\"inject-error-interval\":0}"]
> [2022/06/21 11:18:47.729 +00:00] [INFO] [task_runner.go:219] ["Launching task"] [id=d197b818-5933-4bbf-bb93-89cb533b54fc] [runtime-task-count=1]
> [2022/06/21 11:18:47.729 +00:00] [INFO] [worker.go:568] ["heartbeat pong received"] [master-id=dataflow-engine-job-manager] [msg="{\"send-time\":733284376536,\"reply-time\":\"2022-06-21T11:18:26.621443703Z\",\"to-worker-id\":\"d197b818-5933-4bbf-bb93-89cb533b54fc\",\"epoch\":7,\"is-finished\":false}"]
> [2022/06/21 11:18:50.731 +00:00] [DEBUG] [master_client.go:231] ["sending heartbeat"] [worker=d197b818-5933-4bbf-bb93-89cb533b54fc]
> [2022/06/21 11:18:50.731 +00:00] [INFO] [master_client.go:236] ["sending heartbeat success"] [worker=d197b818-5933-4bbf-bb93-89cb533b54fc] [master-id=dataflow-engine-job-manager]
> [2022/06/21 11:18:50.730 +00:00] [WARN] [center.go:53] ["More than one error is received"] [error="[DFLOW:ErrWorkerSuicide]worker has committed suicide due to master(dataflow-engine-job-manager) having timed out"] [errorVerbose="[DFLOW:ErrWorkerSuicide]worker has committed suicide due to master(dataflow-engine-job-manager) having timed out\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/engine/framework.(*DefaultBaseWorker).runWatchDog\n\tgithub.com/pingcap/tiflow/engine/framework/worker.go:544\ngithub.com/pingcap/tiflow/engine/framework.(*DefaultBaseWorker).startBackgroundTasks.func2\n\tgithub.com/pingcap/tiflow/engine/framework/worker.go:501\nruntime.goexit\n\truntime/asm_amd64.s:1571"]
> ```

It seems it is not caused by this PR. I create an issue for this unstable test. https://github.com/pingcap/tiflow/issues/5969

**Review by maxshuang — APPROVED** [review]

**Pull request merged**